pkg/combination/combination.go

```diff
@@ -51,39 +58,85 @@ func WithArgs(args map[string][]string) StreamOption {
-		cs.solveAhead = true
+		cs.solveAhead = boolVarValue
 	}
 }
 
-// Next gets the next combination from the stream.
-// TODO: Currently this does not solve each combination iteratively and will
-//       need to do so in the future to ensure an optimal use of memory. This
-//       is currently more of a stub to allow consumer packages to maintain its
-// 		 interface.
-func (cs *stream) Next(ctx context.Context) (map[string]string, error) {
-	if cs.solveAhead && !cs.solved {
-		if err := cs.solve(); err != nil {
-			return nil, err
-		}
+// nextIterativeCombination() gets the next combination from the stream.
+// By using this, the stream will solve each combination
+// iteratively.
+func (cs *stream) nextIterativeCombination() (map[string]string, error) {
+	if cs.solved {
+		return nil, nil
 	}
+	// Check to see if anymore combinations exist
 
-	if len(cs.combinations) == 0 {
-		if cs.solved {
-			return nil, nil
-		}
-		return nil, ErrCombinationsNotSolved
+	// Edge case: No parameterListFromArgs
+	if len(cs.parameterListFromArgs) == 0 {
+		cs.solved = true
+		return nil, ErrNoArgsSet
 	}
 
-	next := cs.combinations[0]
-	cs.combinations = cs.combinations[1:]
+	// comboList is a variable that holds a list of combinations to be returned.
+	comboList := map[string]string{}
+
+	// Generate the list of combinations based off current positions
+	for x := 0; x < len(cs.parameterListFromArgs); x++ {
+		combo := cs.args[cs.parameterListFromArgs[x]][cs.positionsMapInArgs[x]]
+		key := cs.parameterListFromArgs[x]
+		comboList[key] = combo
+	}
 
-	return next, nil
+	// Iterates through position map in reverse
+	// looking for the first updatable value then breaks loop.
+	// Otherwise, resets values to zero, we know to update last parameter based off i.
+	var i int
+	for i = len(cs.positionsMapInArgs) - 1; i > cs.nextParameterToUpdate; i-- {
+		if cs.positionsMapInArgs[i]+1 < len(cs.args[cs.parameterListFromArgs[i]]) {
+			cs.positionsMapInArgs[i]++
+			break
+		} else {
+			cs.positionsMapInArgs[i] = 0
+		}
+	}
+
+	// Checks to see if we need to update lastParameter based
+	// off value of i.
+	if i == cs.nextParameterToUpdate {
+		// Checks to see if this is the last argument of the parameter.
+		// Then updates parameter, and checks to see if the combination is solved.
+		if cs.positionsMapInArgs[cs.nextParameterToUpdate]+1 == len(cs.args[cs.parameterListFromArgs[cs.nextParameterToUpdate]]) {
+			cs.positionsMapInArgs[cs.nextParameterToUpdate] = 0
+			cs.nextParameterToUpdate--
+		}
+		// If combination is not solved, we find the next nextParameterToUpdate,
+		// if nextParameterToUpdate has only 1 argument. Also, will mark as solved
+		// if reach end up parameters.
+		runner := true
+		for !cs.solved && runner && cs.nextParameterToUpdate != -1 {
+			if len(cs.args[cs.parameterListFromArgs[cs.nextParameterToUpdate]]) == 1 {
+				cs.nextParameterToUpdate--
+			} else {
+				runner = false
+				cs.positionsMapInArgs[cs.nextParameterToUpdate]++
+			}
```

Comment:
nit/not blocking: you can remove the if/else conditional structure here, which improves readability by using the `continue` keyword:

```go
if len(cs.args[cs.parameterListFromArgs[cs.nextParameterToUpdate]]) == 1 {
	cs.nextParameterToUpdate--
	continue
}
runner = false
cs.positionsMapInArgs[cs.nextParameterToUpdate]++
```